Mathlib/Algebra/AddConstMap/Basic.lean

```diff
@@ -0,0 +1,451 @@
+for bundled maps satisfying `f (x + a) = f x + b`.
+
+We use parameters `a` and `b` instead of `1` to accomodate for two use cases:
+
+- maps between circles of different lengths;
+- self-maps \(f\colon S^1\to  S^1\) of degree other than one,
+  including orientation-reversing maps.
+-/
+
+open Function Set
+
+/-- A bundled map `f : G → H` such that `f (x + a) = f x + b` for all `x`.
+
+One can think about `f` as a lift to `G` of a map between two `AddCircle`s. -/
+structure AddConstMap (G H : Type*) [Add G] [Add H] (a : G) (b : H) where
+  /-- The underlying function of an `AddConstMap`.
+  Use automatic coercion to function instead. -/
+  protected toFun : G → H
+  /-- An `AddConstMap` satisfies `f (x + a) = f x + b`. Use `map_add_const` instead.-/
+  map_add_const' (x : G) : toFun (x + a) = toFun x + b
+
+@[inherit_doc]
+scoped [AddConstMap] notation:25 G " →+c[" a ", " b "] " H => AddConstMap G H a b
+
+/-- Typeclass for maps satisfying `f (x + a) = f x + b`.
+
+Note that `a` and `b` are `outParam`s,
+so one should not add instances like
+`[AddConstMapClass F G H a b] : AddConstMapClass F G H (-a) (-b)`. -/
+class AddConstMapClass (F : Type*) (G H : outParam (Type*)) [Add G] [Add H]
+    (a : outParam G) (b : outParam H) extends FunLike F G fun _ ↦ H where
+  /-- A map of `AddConstMapClass` class semiconjugates shift by `a` to the shift by `b`:
+  `∀ x, f (x + a) = f x + b`. -/
+  map_add_const (f : F) (x : G) : f (x + a) = f x + b
+
+namespace AddConstMapClass
+
+/-!
+### Properties of `AddConstMapClass` maps
+
+In this section we prove properties like `f (x + n • a) = f x + n • b`.
+-/
+
+attribute [simp] map_add_const
+
+variable {F G H : Type*} {a : G} {b : H}
+
+protected theorem semiconj [Add G] [Add H] [AddConstMapClass F G H a b] (f : F) :
+    Semiconj f (· + a) (· + b) :=
+  map_add_const f
+
+@[simp]
+theorem map_add_nsmul [AddMonoid G] [AddMonoid H] [AddConstMapClass F G H a b]
+    (f : F) (x : G) (n : ℕ) : f (x + n • a) = f x + n • b := by
+  simpa using (AddConstMapClass.semiconj f).iterate_right n x
+
+@[simp]
+theorem map_add_nat' [AddMonoidWithOne G] [AddMonoid H] [AddConstMapClass F G H 1 b]
+    (f : F) (x : G) (n : ℕ) : f (x + n) = f x + n • b := by simp [← map_add_nsmul]
+
+theorem map_add_one [AddMonoidWithOne G] [Add H] [AddConstMapClass F G H 1 b]
+    (f : F) (x : G) : f (x + 1) = f x + b := map_add_const f x
+
+@[simp]
+theorem map_add_ofNat' [AddMonoidWithOne G] [AddMonoid H] [AddConstMapClass F G H 1 b]
+    (f : F) (x : G) (n : ℕ) [n.AtLeastTwo] :
+    f (x + no_index (OfNat.ofNat n)) = f x + (OfNat.ofNat n : ℕ) • b :=
+  map_add_nat' f x n
+
+theorem map_add_nat [AddMonoidWithOne G] [AddMonoidWithOne H] [AddConstMapClass F G H 1 1]
+    (f : F) (x : G) (n : ℕ) : f (x + n) = f x + n := by simp
+
+theorem map_add_ofNat [AddMonoidWithOne G] [AddMonoidWithOne H] [AddConstMapClass F G H 1 1]
+    (f : F) (x : G) (n : ℕ) [n.AtLeastTwo] :
+    f (x + no_index (OfNat.ofNat n)) = f x + OfNat.ofNat n := map_add_nat f x n
```

Comment:
I don't think you need the `no_index` here: it is only relevant for simp lemmas. Same thing below.